src/Core/Managed/Shared/ActivityExtensions.cs

```diff
@@ -0,0 +1,55 @@
+﻿#if !NET40
+namespace Microsoft.ApplicationInsights
+{
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Runtime.CompilerServices;
+
+    internal static class ActivityExtensions
+    {
+        private const string OperationNameTag = "OperationName";
+        private static bool isInitialized = false;
+        private static bool isEnabled = false;
+
+        /// <summary>
+        /// Checks that Activity API could be called (DiagnosticSource DLL is loaded).
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoOptimization | MethodImplOptions.NoInlining)] 
+        public static bool IsActivityEnabled()
+        {
+            if (isInitialized)
+            {
+                return isEnabled;
+            }
+
+            try
+            {
+                var activity = new Activity("IsEnabled");
+                isEnabled = true;
+                isInitialized = true;
+                return true;
+            }
+            catch (System.IO.FileNotFoundException)
```

Comment:
TypeLoadException will be thrown if the assembly was loaded, but it does not have the Activity type in it. In our case, we are sure Activity is there (because we insist on DiagnosticSource 4.4.0), but we assume that someone will try to run AppInsights.dll without the DiagnosticSource.dll. In this case assembly will not be found and it will cause FileNotFoundException.

I like the approach with Activator.CreateInstance and I assume it will not be optimized away.